File: SmoochoPOS-Android/app/src/main/java/com/beloop/pos/data/model/Order.kt
package com.beloop.pos.data.model

import android.os.Parcelable
import androidx.room.Entity
import androidx.room.PrimaryKey
import kotlinx.parcelize.Parcelize
import java.util.Date

@Entity(tableName = "orders")
@Parcelize
data class Order(
    @PrimaryKey(autoGenerate = true)
    val id: Long = 0,
    val orderNumber: String,
    val items: List<OrderItem>,
    val subtotal: Double,
    val discount: Double = 0.0,
    val discountType: DiscountType = DiscountType.FLAT,
    val tax: Double,
    val total: Double,
    val paymentMethod: PaymentMethod,
    val paymentStatus: PaymentStatus,
    val orderStatus: OrderStatus = OrderStatus.PENDING,
    val cashierId: Long,
    val customerName: String? = null,
    val customerPhone: String? = null,
    val customerEmail: String? = null,
    val notes: String? = null,
    val tableNumber: String? = null,
    val estimatedTime: Int? = null, // in minutes
    val createdAt: Date = Date(),
    val updatedAt: Date = Date(),
    val completedAt: Date? = null,
    val lastSyncedAt: Date? = null,
    val syncStatus: SyncStatus = SyncStatus.PENDING
) : Parcelable

@Parcelize
data class OrderItem(
    val productId: Long,
    val productName: String,
    val quantity: Int,
    val unitPrice: Double,
    val totalPrice: Double,
    val specialInstructions: String? = null,
    val modifications: List<Modification> = emptyList()
) : Parcelable

@Parcelize
data class Modification(
    val name: String,
    val price: Double = 0.0,
    val isRemoval: Boolean = false
) : Parcelable

enum class DiscountType {
    FLAT, PERCENTAGE
}

enum class PaymentMethod {
    CASH, CARD, UPI, WALLET, CREDIT
}

enum class PaymentStatus {
    PENDING, COMPLETED, FAILED, REFUNDED
}

enum class OrderStatus {
    PENDING, CONFIRMED, PREPARING, READY, SERVED, COMPLETED, CANCELLED
}

enum class SyncStatus {
    PENDING, SYNCED, FAILED, CONFLICT
}

File: src/domain/entities/Order.ts
import { CartItem } from './CartItem'
import { ProductId, ProductName, ProductPrice } from './Product'

export class OrderId {
  constructor(public readonly value: number) {}
}

export class OrderNumber {
  constructor(public readonly value: string) {}
}

export interface OrderItem {
  productId: ProductId
  productName: ProductName
  quantity: number
  price: ProductPrice
  total: number
}

export type PaymentMethod = 'cash' | 'card' | 'upi'
export type PaymentStatus = 'pending' | 'completed' | 'failed'
export type OrderType = 'takeaway' | 'delivery' | 'dine-in'

export interface OrderMoney {
  value: number
}

export class Order {
  constructor(
    public readonly id: OrderId,
    public readonly orderNumber: OrderNumber,
    public readonly items: OrderItem[],
    public readonly subtotal: OrderMoney,
    public readonly discount: OrderMoney,
    public readonly discountType: 'flat' | 'percentage',
    public readonly tax: OrderMoney,
    public readonly total: OrderMoney,
    public readonly paymentMethod: PaymentMethod,
    public readonly paymentStatus: PaymentStatus,
    public readonly orderType: OrderType,
    public readonly customerName?: string,
    public readonly customerPhone?: string,
    public readonly notes?: string,
    public readonly cashierId: number = 1,
    public readonly createdAt: Date = new Date(),
    public readonly updatedAt: Date = new Date()
  ) {}

  // Business logic methods
  isCompleted(): boolean {
    return this.paymentStatus === 'completed'
  }

  isPending(): boolean {
    return this.paymentStatus === 'pending'
  }

  isFailed(): boolean {
    return this.paymentStatus === 'failed'
  }

  getItemCount(): number {
    return this.items.reduce((count, item) => count + item.quantity, 0)
  }

  hasCustomerInfo(): boolean {
    return !!(this.customerName || this.customerPhone)
  }

  updatePaymentStatus(status: PaymentStatus): Order {
    return new Order(
      this.id,
      this.orderNumber,
      this.items,
      this.subtotal,
      this.discount,
      this.discountType,
      this.tax,
      this.total,
      this.paymentMethod,
      status,
      this.orderType,
      this.customerName,
      this.customerPhone,
      this.notes,
      this.cashierId,
      this.createdAt,
      new Date()
    )
  }

  static fromCartItems(
    cartItems: CartItem[],
    orderNumber: OrderNumber,
    paymentMethod: PaymentMethod,
    orderType: OrderType,
    discount: OrderMoney = { value: 0 },
    discountType: 'flat' | 'percentage' = 'flat',
    taxRate: number = 0,
    deliveryCharge: number = 0,
    customerName?: string,
    customerPhone?: string
  ): Order {
    const items: OrderItem[] = cartItems.map(item => ({
      productId: item.product.id,
      productName: item.product.name,
      quantity: item.quantity.value,
      price: item.product.price,
      total: item.getTotalPrice()
    }))

    const subtotal = { value: cartItems.reduce((sum, item) => sum + item.getTotalPrice(), 0) }
    const discountAmount = discountType === 'flat' 
      ? Math.min(discount.value, subtotal.value)
      : (subtotal.value * discount.value) / 100
    
    const taxableAmount = subtotal.value - discountAmount
    const tax = { value: (taxableAmount * taxRate) / 100 }
    const total = { value: taxableAmount + tax.value + deliveryCharge }

    return new Order(
      { value: 0 }, // Will be set by repository
      orderNumber,
      items,
      subtotal,
      { value: discountAmount },
      discountType,
      tax,
      total,
      paymentMethod,
      'completed',
      orderType,
      customerName,
      customerPhone,
      `Order Type: ${orderType.toUpperCase()}${deliveryCharge > 0 ? ` | Delivery: ₹${deliveryCharge}` : ''}`
    )
  }
}
